File: src/Authzi.MicrosoftOrleans.Interoperability/Authorization/OutgoingGrainCallAuthorizationFilter.cs
using System.Threading.Tasks;
using Authzi.Security;
using Authzi.Security.AccessToken;
using Authzi.Security.Authorization;
using Microsoft.Extensions.Logging;
using Orleans;

namespace Authzi.MicrosoftOrleans.Authorization
{
    public class OutgoingGrainCallAuthorizationFilter : GrainAuthorizationFilterBase, IOutgoingGrainCallFilter
    {
        public OutgoingGrainCallAuthorizationFilter(IAccessTokenVerifier accessTokenVerifier, 
            IAuthorizationExecutor authorizeHandler,
            // ReSharper disable once SuggestBaseTypeForParameter
            ILogger<OutgoingGrainCallAuthorizationFilter> logger) : base(accessTokenVerifier, authorizeHandler)
        {
            Logger = logger;
        }


        public async Task Invoke(IOutgoingGrainCallContext context)
        {
            if (AuthorizationAdmission.IsRequired(context))
            {
                await AuthorizeAsync(context);

                var grainType = context.Grain.GetType();
                Log(LoggingEvents.OutgoingGrainCallAuthorizationPassed,
                    grainType.Name, context.InterfaceMethod.Name);
            }

            await context.Invoke();
        }
    }
}

File: src/AuthZI.MicrosoftOrleans/OutgoingGrainCallAuthorizationFilter.fs
namespace AuthZI.MicrosoftOrleans.Authorization

open AuthZI.Security.AccessToken
open AuthZI.Security.Authorization
open AuthZI.Security;
open Microsoft.Extensions.Logging
open Orleans
open System.Threading.Tasks

//type OutgoingGrainCallAuthorizationFilter(accessTokenVerifier: IAccessTokenVerifier,
//    authorizeHandler: IAuthorizationExecutor, logger: ILogger<OutgoingGrainCallAuthorizationFilter>) as this =

//    //member _.AuthorizeAsync(context) = base.AuthorizeAsync(context)
//    //member _.Log(eventId, grainTypeName, interfaceMethodName) = base.Log(eventId, grainTypeName, interfaceMethodName)
//    interface IOutgoingGrainCallFilter with
//        member _.Invoke(context: IOutgoingGrainCallContext) =
//            task {
//                if AuthorizationDeterminer.IsRequired context.InterfaceMethod then
//                    //let! claims = this.AuthorizeAsync(context)
//                    //let grainType = context.Grain.GetType()

//                    //this.Log(LoggingEvents.OutgoingGrainCallAuthorizationPassed,
//                    //    grainType.Name, context.InterfaceMethod.Name)

//                do! context.Invoke()
//            } :> Task